src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs

```diff
@@ -132,102 +154,165 @@ public virtual byte[] GetValue()
-                }
-                else if (controls[i].Type == "1.2.840.113556.1.4.1504")
-                {
-                    // The control is an AsqControl.
-                    object[] o = BerConverter.Decode("{e}", value);
-                    Debug.Assert((o != null) && (o.Length == 1));
+                    if (controls[i].Type == "1.2.840.113556.1.4.319")
+                    {
+                        // The control is a PageControl, as described in RFC 2696.
+                        byte[] cookie = Array.Empty<byte>();
 
-                    int result = (int)o[0];
-                    AsqResponseControl asq = new AsqResponseControl(result, controls[i].IsCritical, controls[i].GetValue());
-                    controls[i] = asq;
-                }
-                else if (controls[i].Type == "1.2.840.113556.1.4.841")
-                {
-                    // The control is a DirSyncControl.
-                    object[] o = BerConverter.Decode("{iiO}", value);
-                    Debug.Assert(o != null && o.Length == 3);
+                        AsnDecoder.ReadSequence(asnSpan, AsnEncodingRules.BER, out int sequenceContentOffset, out int sequenceContentLength, out _);
+                        ThrowUnless(sequenceContentLength > 0);
 
-                    int moreData = (int)o[0];
-                    int count = (int)o[1];
-                    byte[] dirsyncCookie = (byte[])o[2];
+                        asnSpan = asnSpan.Slice(sequenceContentOffset, sequenceContentLength);
+                        asnReadSuccessful = AsnDecoder.TryReadInt32(asnSpan, AsnEncodingRules.BER, out int size, out int bytesConsumed);
+                        ThrowUnless(asnReadSuccessful);
+                        asnSpan = asnSpan.Slice(bytesConsumed);
 
-                    DirSyncResponseControl dirsync = new DirSyncResponseControl(dirsyncCookie, (moreData == 0 ? false : true), count, controls[i].IsCritical, controls[i].GetValue());
-                    controls[i] = dirsync;
-                }
-                else if (controls[i].Type == "1.2.840.113556.1.4.474")
-                {
-                    // The control is a SortControl.
-                    int result;
-                    string attribute = null;
-                    object[] o = BerConverter.TryDecode("{ea}", value, out bool decodeSucceeded);
+                        // If present, the remaining bytes in the control are expected to be an octet string.
+                        if (!asnSpan.IsEmpty)
+                        {
+                            // user expects cookie with length 0 as paged search is done. In this situation, there'll still be two bytes
+                            // for the ASN.1 tag.
+                            cookie = AsnDecoder.ReadOctetString(asnSpan, AsnEncodingRules.BER, out bytesConsumed);
+                            asnSpan = asnSpan.Slice(bytesConsumed);
+                        }
+                        ThrowUnless(asnSpan.IsEmpty);
```

Comment:
I'm a fan of this check (it matches what the RFC says), but it's not something that BerConverter.Decode does, so it's introducing an exception where there wasn't one previously.

That's not forbidden, but there should be a test for it (maybe there already is) and it needs to be part of a breaking change doc.

Reply:
There doesn't seem to be any self-contained test coverage of the controls decoded in this method; the closest it could be said to come is the processing which takes place when somebody tests against a real LDAP server. I'll add this coverage, then broaden it to explicitly test trailing data in SEQUENCEs.